### PR TITLE
Connie/threading proposal

### PR DESCRIPTION
**Pull request by Connicpu**
Implement #35 

The implementation of wait/notify is far from optimized but it works for now. It can be greatly improved if the spec relaxes the requirement for no spurious failures.

imports.wast will not be able to pass until multi-memory is implemented

### CODE REVIEW

File: innative-env/wait_list.c [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+  }
+
+  in_wait_list* value = map->entries[idx].value;
+  in_atomic_incr(&value->refs);
+  return value;
+}
+
+in_wait_list* _innative_internal_env_wait_map_get(in_wait_map* map, void* address, int create)
+{
+  in_wait_list* result;
```
Comment (ErikMcClure) [inline innative-env/wait_list.c:167]: Can this be declared at the moment result is assigned instead of up here?

File: innative/lexer.cpp [hunk 2]
```diff
@@ -534,7 +551,7 @@ void innative::TokenizeWAT(Queue<WatToken>& tokens, const char* s, const char* e
           tokens.Push(WatToken{ kh_val(tokenhash, iter), begin, line, column });
         else
         {
-          uint8_t op = GetInstruction(ref);
+          uint16_t op = GetInstruction(ref);
```
Comment (ErikMcClure) [inline innative/lexer.cpp:554]: Why was this changed to uint16_t? Is it capturing all possible instruction bytes? Does that work properly? If we are trying to capture all possible instruction bytes, maybe this type should be defined somewhere else so it can be derived from the current max instruction size.

File: innative-env/wait_list.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+// TODO
+#endif
+
+struct in_wait_entry
+{
+  in_platform_condvar condvar;
+  union
+  {
+    int32_t signaled;
+    in_wait_entry* next_free_node;
```
Comment (ErikMcClure) [inline innative-env/wait_list.h:67]: Not using `struct in_wait_entry` will cause problems on more strict C compilers (like GCC). This header is intended to be accessible from C, so all instaces of struct types should be preceded by `struct`.

File: innative/atomic_instruction_details.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    Xor,
+    Xchg,
+    CmpXchg,
+
+    INVALID_CAT,
+  };
+
+  constexpr OpGroup GetOpGroup(uint8_t op) { return OpGroup((op - OP_START) / OP_GROUP_SIZE); }
+  constexpr int GetOpType(uint8_t op) { return (op - OP_START) % OP_GROUP_SIZE; }
+  constexpr uint8_t IsI64(int opType) { return uint8_t(0x01'01'01'00'00'01'00 >> (opType * 8)); }
```
Comment (ErikMcClure) [inline innative/atomic_instruction_details.h:27]: Can a comment be left here explaining what the heck this is doing, and what the byte separations in the constant are for?

File: innative/atomic_instructions.cpp [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+{
+  IN_ERROR err;
+
+  llvmVal* base;
+  if(err = PopType(TE_i32, base))
+    return err;
+
+  varuint32 memflags = ins.immediates[0]._varuint32;
+  varuint32 offset   = ins.immediates[1]._varuint32;
+  varuint32 memory   = ins.immediates[2]._varuint32;
```
Comment (ErikMcClure) [inline innative/atomic_instructions.cpp:43]: I am unable to verify in the spec that this is the expected type of the memory index. Do you have a link or an example of a test that shows it must be considered a 32-bit integer?

File: innative/atomic_instructions.cpp [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+    llvmVal* loaded = builder.CreateExtractValue(cmpxchg, 0);
+    if(varsize > align)
+      loaded = builder.CreateZExt(loaded, varty);
+
+    return PushReturn(loaded);
+  });
+}
+
+using namespace innative::atomic_details;
```
Comment (ErikMcClure) [inline innative/atomic_instructions.cpp:214]: `innative::` here and the other `innative::` prefixes could probably be removed by just adding `using namespace innative` at the top of the file, unless this creates a conflict.

File: innative/atomic_instructions.cpp [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+
+IN_ERROR innative::Compiler::CompileAtomicInstruction(Instruction& ins)
+{
+  using RmwOp = llvm::AtomicRMWInst::BinOp;
```
Comment (ErikMcClure) [inline innative/atomic_instructions.cpp:218]: I assume this is scoped to just this function, is that correct?

File: innative/compile.h [hunk 8]
```diff
@@ -110,7 +113,7 @@ namespace innative {
     void PolymorphicStack();
     llvmVal* GetMemSize(llvm::GlobalVariable* target);
     varuint32 GetFirstType(varuint32 type);
-    llvmVal* GetMemPointer(llvmVal* base, llvm::PointerType* pointer_type, varuint7 memory, varuint32 offset);
+    llvmVal* GetMemPointer(llvmVal* base, llvm::PointerType* pointer_type, varuint32 memory, varuint32 offset);
```
Comment (ErikMcClure) [inline innative/compile.h:116]: As stated above, please double check that memory indices were actually changed to 32-bit integers.

File: innative/constants.cpp [hunk 9]
```diff
@@ -267,9 +271,12 @@ namespace innative {
       { ERR_INVALID_LIMITS, "size minimum must not be greater than maximum" },
       { ERR_MEMORY_MAXIMUM_TOO_LARGE, "memory size must be at most 65536 pages (4GiB)" },
       { ERR_MEMORY_MINIMUM_TOO_LARGE, "memory size must be at most 65536 pages (4GiB)" },
+      { ERR_SHARED_MEMORY_MAXIMUM_MISSING, "shared memory must have maximum" },
       { ERR_INVALID_INITIALIZER, "constant expression required" },
       { ERR_EMPTY_IMPORT, "unknown import" },
       { ERR_WAT_PARAM_AFTER_RESULT, "unexpected token" },
+
```
Comment (ErikMcClure) [inline innative/constants.cpp:278]: remove extraneous space

File: innative/parse.cpp [hunk 10]
```diff
@@ -150,12 +151,13 @@ IN_ERROR innative::ParseFunctionType(Stream& s, FunctionType& sig, const Environ
 
 IN_ERROR innative::ParseResizableLimits(Stream& s, ResizableLimits& limits)
 {
+  // This will only actually read a single byte
```
Comment (ErikMcClure) [inline innative/parse.cpp:154]: Why is this using ParseVarUInt32 if it will only actually read a single byte? If there is only one byte it should be varuint7, unless the spec has done something very strange.

File: innative/parse.cpp [hunk 11]
```diff
@@ -331,11 +335,20 @@ IN_ERROR innative::ParseInstruction(Stream& s, Instruction& ins, const Environme
...
   case OP_i64_store32:
+    // Alignment and memidx bit flag
     ins.immediates[0]._varuint32 = s.ReadVarUInt32(err);
 
+    // Offset
     if(err >= 0)
       ins.immediates[1]._varuptr = s.ReadVarUInt32(err); // Currently 32-bit because all memories are 32-bit
 
+    // If bit 6 is set, read a memidx value
+    if(err >= 0 && (ins.immediates[0]._varuint32 & 0b1000000))
```
Comment (ErikMcClure) [inline innative/parse.cpp:346]: Should this access really be _varuint32? Please check or confirm the types of memidx values.

File: innative/parse.cpp [hunk 12]
```diff
@@ -951,3 +969,46 @@ IN_ERROR innative::ParseExportFixup(Module& m, ValidationError*& errors, const E
...
+  err = ParseByte(s, ins.opcode[1]);
+  if(err < 0)
+    return err;
+
+  uint8_t alignByte;
+  err = ParseByte(s, alignByte);
+  if(err < 0)
+    return err;
+
+  ins.immediates[0]._varuint32 = alignByte & 0b111;
```
Comment (ErikMcClure) [inline innative/parse.cpp:988]: Why are we assigning a single byte to _varuint32?

File: innative/serialize.cpp [hunk 13]
```diff
@@ -351,6 +361,9 @@ void Serializer::TokenizeModule(bool emitdebug)
     t.Push(WatToken{ WatTokens::INTEGER, 0, 0, 0, limits.minimum });
     if(limits.flags & WASM_LIMIT_HAS_MAXIMUM)
       t.Push(WatToken{ WatTokens::INTEGER, 0, 0, 0, limits.maximum });
+    if(limits.flags & WASM_LIMIT_SHARED)
+      t.Push(WatToken{ WatTokens::SHARED });
+    // TODO: Should we serialize an UNSHARED token in the other case?
```
Comment (ErikMcClure) [inline innative/serialize.cpp:366]: No, unless other wasm serializers decide to do this (but none of them support threads yet).

File: innative/utility.cpp [hunk 14]
```diff
@@ -136,7 +136,7 @@ namespace innative {
       return h;
     }
 
-    uint8_t GetInstruction(StringSpan ref)
+    uint16_t GetInstruction(StringSpan ref)
```
Comment (ErikMcClure) [inline innative/utility.cpp:139]: Once again, if this is supposed to capture all instruction bytes in a single type, this type should be defined based off of the maximum number of instruction bytes.

File: innative/lexer.cpp (hunk 2 above)
Comment (Connicpu) [reply innative/lexer.cpp:554]: Yeah, the hashmap that GetInstruction reads from already contains the 2-byte versions of all the opcodes, with opcode[0] in the lower 8 bits and the high 8 bits set to 0 for single byte instructions. Using a typedef might be a good idea, not sure what kind of refactoring that'd require but I can go through it

File: innative-env/wait_list.h (hunk 3 above)
Comment (Connicpu) [reply innative-env/wait_list.h:67]: All of those structs are typedef'd above on line 21^^

File: innative/atomic_instructions.cpp (hunk 6 above)
Comment (Connicpu) [reply innative/atomic_instructions.cpp:214]: Oh yep, those were all redundant heh. I never noticed but visual studio just inserts that when you use the "Copy signature to the clipboard" helper from the header

File: innative/atomic_instructions.cpp (hunk 7 above)
Comment (Connicpu) [reply innative/atomic_instructions.cpp:218]: Yeah, this is just scoped to the function to cut down on the noise for accessing that enum

File: innative/parse.cpp (hunk 10 above)
Comment (Connicpu) [reply innative/parse.cpp:154]: I'm actually not sure on this one tbh. Technically according to the spec the only valid values here are {0,1,3}, but it was already reading a varuint32. I can change it to varint7 if you'd prefer though

File: innative/parse.cpp (hunk 12 above)
Comment (Connicpu) [reply innative/parse.cpp:988]: This was just for symmetry with the existing load/store instructions. They also used the _varuint32 field even though the spec says they can't be bigger than 0x7
